_posts/2016-09-10-arm64-icache.md

```diff
@@ -0,0 +1,88 @@
+and the `SIGILL` would [always happen to be somewhere between](https://gist.github.com/lewurm/97dff0a56929b56a0fc5ab49af06fd06) `0x40-0x7f` or `0xc0-0xff`.
+We aligned the memory dump to help verifying whether the code allocator was doing something funky.
+
+```
+$ grep SIGILL *.log
+custom_01.log:E/mono           (13964): SIGILL at ip=0x0000007f4f15e8d0
+custom_02.log:E/mono           (13088): SIGILL at ip=0x0000007f8ff76cc0
+custom_03.log:E/mono           (12824): SIGILL at ip=0x0000007f68e93c70
+custom_04.log:E/mono           (12876): SIGILL at ip=0x0000007f4b3d55f0
+custom_05.log:E/mono           (13008): SIGILL at ip=0x0000007f8df1e8d0
+custom_06.log:E/mono           (14093): SIGILL at ip=0x0000007f6c21edf0
+[...]
+```
+
+With that we came with our first good hypothesis. Bad cache flushing was happening only on the upper 64 bytes of every 128 bytes block.
+Those numbers, if you deal with low level programming, immediately reminds you of cache line sizes. And that is where it all started to make
+sense.
+
+Here is a pseudo version of how `libgcc` [does cache flushing on arm64](https://android.googlesource.com/toolchain/gcc/+/master/gcc-4.9/libgcc/config/aarch64/sync-cache.c#54):
+
+```c
+void __clear_cache (char *address, size_t size)
+{
+	static int cache_line_size = 0;
+	if (!cache_line_size)
+		cache_line_size = get_current_cpu_cache_line_size ();
+
+	for (int i = 0; i < size; i += cache_line_size)
+		flush_cache_line (address + i);
+}
+```
+
+In the above pseudo-code `get_current_cpu_cache_line_size` is a CPU instruction that returns the line size of its caches and `flush_cache_line`
+flushes the cache line that contains the supplied address.
+
+At that point we were using our own version of that function, so we instrumented it to print the cache line size as returned by the CPU and, lo and behold,
+it printed both 128 and 64. We double verified that this was indeed the case. So we went to see that particular CPU manual and it turns out that the big core has
+a 128 bytes cache line but on the LITTLE core it is only 64 bytes for the instruction cache.
+
+So what was happening is that `__clear_cache` would be called first on a big core and caches 128 as the instruction cache line size. Later it would be called on one
+of the LITTLE cores and would skip every other cache line when flushing. It doesn't get simpler than that. We removed the caching and it all worked.
+
+# Summary
+
+Some ARM big.LITTLE CPUs can have cores with different cache line sizes and pretty much no code out there is ready to deal with it as they
+assume all cores to be symmetrical.
+
+Worse than that, not even the ARM ISA is ready for that. An astute reader might realize that computing the cache line on every invocation
+is not enough for user space code as it can be rescheduled right after on a core with the smaller cache line size. As a workaround we detect
```

Comment:
It depends a lot on how much cache shoot down a core migration triggers.

It would be an issue if the kernel only makes the 2 cores coherent and not all
source core caches globally coherent.